core/src/main/java/org/bitcoinj/core/Transaction.java

```diff
@@ -87,29 +87,25 @@
  * <p>Instances of this class are not safe for use by multiple threads.</p>
  */
 public class Transaction extends ChildMessage {
+    private static final Comparator<Transaction> SORT_TX_BY_ID = Comparator.comparing(Transaction::getTxId);
     /**
      * A comparator that can be used to sort transactions by their updateTime field. The ordering goes from most recent
      * into the past.
      */
-    public static final Comparator<Transaction> SORT_TX_BY_UPDATE_TIME = (tx1, tx2) -> {
-        Instant t1 = tx1.getUpdateTimeInstant().orElse(Instant.EPOCH);
-        Instant t2 = tx2.getUpdateTimeInstant().orElse(Instant.EPOCH);
-        final int updateTimeComparison = -(t1.compareTo(t2));
-        //If time1==time2, compare by tx hash to make comparator consistent with equals
-        return updateTimeComparison != 0 ? updateTimeComparison : tx1.getTxId().compareTo(tx2.getTxId());
-    };
+    public static final Comparator<Transaction> SORT_TX_BY_UPDATE_TIME =
+            Comparator.<Transaction, Instant>comparing(
+                            tx -> tx.getUpdateTimeInstant().orElse(Instant.EPOCH),
+                            Comparator.reverseOrder())
+                    .thenComparing(SORT_TX_BY_ID);
     /** A comparator that can be used to sort transactions by their chain height. */
-    public static final Comparator<Transaction> SORT_TX_BY_HEIGHT = (tx1, tx2) -> {
-        final TransactionConfidence confidence1 = tx1.getConfidence();
-        final int height1 = confidence1.getConfidenceType() == ConfidenceType.BUILDING
-                ? confidence1.getAppearedAtChainHeight() : Block.BLOCK_HEIGHT_UNKNOWN;
-        final TransactionConfidence confidence2 = tx2.getConfidence();
-        final int height2 = confidence2.getConfidenceType() == ConfidenceType.BUILDING
-                ? confidence2.getAppearedAtChainHeight() : Block.BLOCK_HEIGHT_UNKNOWN;
-        final int heightComparison = -(Integer.compare(height1, height2));
-        //If height1==height2, compare by tx hash to make comparator consistent with equals
-        return heightComparison != 0 ? heightComparison : tx1.getTxId().compareTo(tx2.getTxId());
-    };
+    public static final Comparator<Transaction> SORT_TX_BY_HEIGHT =
+            Comparator.<Transaction, Integer>comparing(
+                            tx -> tx.getConfidence().getConfidenceType() == ConfidenceType.BUILDING ?
+                                    tx.getConfidence().getAppearedAtChainHeight() :
+                                    Block.BLOCK_HEIGHT_UNKNOWN,
+                            Comparator.reverseOrder())
+                    .thenComparing(SORT_TX_BY_ID);
+
```

Comment:
Here the extracted method helps by:
1.  Caches the `TransactionConfidence` in a local variable
2. Eliminates the need for the `<Transaction, Integer>`
3. Makes the compactor more readable

Again, you might want a better method name. And it also would be possible to add this as an instance method.

```
    public static final Comparator<Transaction> SORT_TX_BY_HEIGHT =
            Comparator.comparing(Transaction::txGetHeight, Comparator.reverseOrder())
                        .thenComparing(SORT_TX_BY_ID);

    private static int txGetHeight(Transaction tx) {
        TransactionConfidence conf = tx.getConfidence();
        return conf.getConfidenceType() == ConfidenceType.BUILDING
                ? conf.getAppearedAtChainHeight()
                : Block.BLOCK_HEIGHT_UNKNOWN;
    }
```

Reply:
The helper method also lets you write a comment, which I lazily didn't do.